### PR TITLE
Config panel

### PR DESCRIPTION
**Pull request by nathanael-h**
## Problem

- *Description of why you made this PR*

## Solution

- *And how do you fix that problem*

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by MayeulC** [comment]
Thank you for starting this! For me, the most important feature would be to add/remove people from bridge admin, bridge users, and relay users.

**Comment by nathanael-h** [comment]
Hello, I will try to understand how this config panels work :exploding_head:  . Managing users allowed to use the bridge would be the 2nd step :two: :slightly_smiling_face: 

For me config is also triggered at install time.

**Comment by nathanael-h** [comment]
Hello, actually I have this panel: 
image

It uses the existing app setting key `encryption` see https://github.com/YunoHost-Apps/mautrix_signal_ynh/blob/master/scripts/install#L70

It does update the `/opt/yunohost/mautrix_signal/config.yaml`, but not at the right line. It adds a "0" or a "1" next to "encryption" instead of next to "allow": 
```yaml
    # See https://docs.mau.fi/bridges/general/end-to-bridge-encryption.html for more info.
    encryption:  0
        # Allow encryption, work in group chat rooms with e2ee enabled
        allow: true
``` 
instead of 
```yaml
    # See https://docs.mau.fi/bridges/general/end-to-bridge-encryption.html for more info.
    encryption:  
        # Allow encryption, work in group chat rooms with e2ee enabled
        allow: 0
``` 
This is the next step. 
After I will have to replace 1/0 by true/false and handle the service restart only if the config was modified. I'd welcome any help :+1:

File: scripts/config [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+source _common.sh
+source /usr/share/yunohost/helpers
+
+ynh_abort_if_errors
+
+#=================================================
+# RETRIEVE ARGUMENTS
+#=================================================
+
+encryption_allowed=$(ynh_app_setting_get --app=$app --key=encryption_allowed)
```
Comment (fflorent) [inline scripts/config:18]: As I understand what is stated in the example `config_panel.toml`: https://github.com/YunoHost/example_ynh/blob/master/config_panel.toml.example#L162-L165

```toml
        ## Note that this feature only works with relatively simple cases
        ## such as `KEY: VALUE`, but won't properly work with
        ## complex data structures like multilin array/lists or dictionnaries.
        ## It also doesn't work with XML format, custom config function call, php define(), ...
```

You should go with a getter and a setter in the `script/config` panel. If I were you, for the setter, I'd go for a simple:

```bash
set__encryption_allowed() {
    ynh_app_setting_set --app=$app --key=encryption_allowed --value=$encryption_allowed
    ynh_add_config --template="../conf/config.yaml" --destination="$final_path/config.yaml"

    chmod 400 "$final_path/config.yaml"
    chown $app:$app "$final_path/config.yaml"
}
```

File: config_panel.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+version = "1.0"
+
+[bind]
+name = "Mautrix Signal configuration"
+
+    [bind.encryption]
+    name = "Manage encryption"
+
+        [bind.encryption.encryption]
```
Comment (fflorent) [inline config_panel.toml:9]: Shouldn't this be `bind.encryption.encryption_allowed` to be more meaningful?

Also this should reuse the name while using `ynh_app_setting_set`.

File: config_panel.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+version = "1.0"
+
+[bind]
```
Comment (fflorent) [inline config_panel.toml:3]: What you declare here is an ID for the `"Mautrix Signal configuration"` panel. What I understood of the config_panel.toml file is that the first level ID is only meant to be meaningful when using it in the CLI.

You may use a different ID if you want to. Like `[main]` or `[bridge]` (and consequently adapt the IDs below).

**Review by fflorent — COMMENTED** [review]
With the hope to help ^^

**Comment by Gredin67** [comment]
Did you check the PR here ? https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/74

File: config_panel.toml (hunk 3 above)
Comment (nathanael-h) [reply config_panel.toml:3]: Thanks for the explanation, I better understood.

File: scripts/config (hunk 1 above)
Comment (nathanael-h) [reply scripts/config:18]: Thanks, this helped me to understand. I'll fix this and use same structure as in mautrix_whatsapp_ynh

**Comment by nathanael-h** [comment]
@Gredin67 Thanks for mentioning this PR, I did not saw it and it will help a lot !!! I started to use the same structure for signal bridge, going step by step in config_panel.

**Comment by Gredin67** [comment]
I tested the upgrade and I can use the config panel to change encryption ! :dancers: 
Would be nice to add at least the user and admin permissions management, as it is implemented, tested and working in https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/74
!testme
What do you think @zamentur @tituspijean should we merge?

**Comment by Gredin67** [comment]
@nathanael-h I've "synced" the changes from mautrix_whatsapp. Both mautrix_whatsapp and mautrix_signal require testing
!testme

**Comment by Gredin67** [comment]
!testme

**Comment by Gredin67** [comment]
after upgrade, I'm getting `Config panel question 'print_level' should be initialized with a value during install or upgrade.` although it is initialized... https://github.com/nathanael-h/mautrix_signal_ynh/blob/config_panel/scripts/upgrade#L159
```
        [appservice.config.print_level]
        ask = "Root logging print level"
        type = "select"
        choices = ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "NOTSET"]
        help = "NOTSET means that all messages will be logged: https://docs.python.org/3.6/library/logging.config.html#configuration-file-format"
        bind = ":__FINALPATH__/config.yaml"
```

**Comment by Gredin67** [comment]
we still need to correct some bind when it does not correspond to setting name https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/74/commits/f151a35cda9a2c4bf1d15136f3f7cadb4802d1d9

**Comment by nathanael-h** [comment]
Hello @Gredin67 thanks a lot for your help on this PR !! As I got the same error saying that `print_level` wasn't initialized, I tried to remove this key from `config_panel.toml` file, but then I got the same error for `appserviceid`, `botnmae`, `signalbot`, and I did not tried further more. Very strange, I keep trying to understand.

**Comment by nathanael-h** [comment]
The two issues might be linked. This seems to help : 
```diff
diff --git a/config_panel.toml b/config_panel.toml
index b411670..e9455cc 100644
--- a/config_panel.toml
+++ b/config_panel.toml
@@ -20,7 +20,7 @@ services = ["__APP__"]
         type = "select"
         choices = ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL", "NOTSET"]
         help = "NOTSET means that all messages will be logged: https://docs.python.org/3.6/library/logging.config.html#configuration-file-format"
-        bind = ":__FINALPATH__/config.yaml"
+        bind = "logging>root>level:__FINALPATH__/config.yaml"
 
     [appservice.bot]
     name = "Robot Settings"
``` 
What do you think?

**Comment by Gredin67** [comment]
I corrected wrong binds, this should work.

And tried something about storing the values for permission here https://github.com/YunoHost-Apps/mautrix_signal_ynh/tree/config_panel_permissions
I'm 100% sure this will not work :smile:

**Comment by Gredin67** [comment]
Now I think it should work https://github.com/YunoHost-Apps/mautrix_signal_ynh/pull/76
We only need to adapt `set__role` and `get__listadmin`, etc.
Don't know how to proceed to merge #76 here. @nathanael-h  feel free to cherry pick in your branch or so.

**Comment by MayeulC** [comment]
@Gredin67 you should probably make your PR #76 to https://github.com/nathanael-h/mautrix_signal_ynh/pulls :)

**Comment by Gredin67** [comment]
Very strange, I've changed the syntax in the config.yaml file to fit the one in mautrix_whatsapp.
From the config_panel/permissions:
- I can set the permissions and they show up like this
```
    permissions:
        '*': relay
        '@gauthier67:sans-nuage.fr': admin
        sans-nuage.fr: user
```
- But permissions are not displayed

I'm wondering if it's an upgrade issue, can someone test https://github.com/nathanael-h/mautrix_signal_ynh/pull/2

**Comment by MayeulC** [comment]
@Gredin67, I don't really like the approach taken by the ynh packages for permissions:

```yaml
permissions:
        '__LISTRELAY__': relay
        '__LISTUSER__': user
        '__LISTADMIN__': admin
```
Those aren't really lists; this section should probably be constructed dynamically, and deduplicate entries, since permissions for Admins > Users > Relay (I think? Users and Relay could be independent).

If you want to do everything in bash, you can use bash dictionary arrays, but the limitted character range for indexing makes them hard to use:

```bash
declare -A permissions

for user in "${adminlist[@]}"
do
  userenc=$(echo "user"|base32 | tr = 9)
  permissions[userenc]="admin"
done

for user in "${userlist[@]}"
do
  userenc=$(echo "$user"|base32 | tr = 9)
  if [[ ! ${permissions[$userenc]+_} ]] # user doesn't exist
  then
    permissions[$userenc]="user"
  fi
done

for user in "${relaylist[@]}"
do
  userenc=$(echo "$user"|base32 | tr = 9)
  if [[ ! ${permissions[$userenc]+_} ]] # user doesn't exist
  then
    permissions[$userenc]="relay"
  fi
done

permissionstr="permissions:"
for userenc in "${!permissions[@]}"
do
  user=$(echo $userenc|tr 9 =|base32 -d)
  permissionstr=$(printf "%s\n    '%s': %s" "$permissionstr" "$user" "${permissions[$userenc]})
done
```

At this point, you have a correct permission string, but what to do? Store it in yunohost's app config file? It would work, but then it will be hard to read it again and parse it into lists. Maybe save both `permissionstr` and the individual lists, and only rely on `permissionstr` for the template.

I feel like setting management in yunohost is extremely broken, which is a shame. These scripts are getting very complex already, so I feel like there is little value in keeping them written in bash and not, say, Python.

**Comment by Gredin67** [comment]
Feel free to contribute to yunohost core, there are plenty of issues open regarding config panel improvement. At least now we have one.

Improvements in yunohost are always a tradeoff between keeping backward compatibility and innovation, it's not that simple ton solve.

And feel free to contribute to this PR, or open a new one to improve it.

Le 11 janvier 2023 18:32:50 GMT+01:00, Mayeul Cantan ***@***.***> a écrit :
>@Gredin67, I don't really like the approach taken by the ynh packages for permissions:
>
>```yaml
>permissions:
>        '__LISTRELAY__': relay
>        '__LISTUSER__': user
>        '__LISTADMIN__': admin
>```
>Those aren't really lists; this section should probably be constructed dynamically, and deduplicate entries, since permissions for Admins > Users > Relay (I think? Users and Relay could be independent).
>
>If you want to do everything in bash, you can use bash dictionary arrays, but the limitted character range for indexing makes them hard to use:
>
>```bash
>declare -A permissions
>
>for user in "${adminlist[@]}"
>do
>  userenc=$(echo "user"|base32 | tr = 9)
>  permissions[userenc]="admin"
>done
>
>for user in "${userlist[@]}"
>do
>  userenc=$(echo "$user"|base32 | tr = 9)
>  if [[ ! ${permissions[$userenc]+_} ]] # user doesn't exist
>  then
>    permissions[$userenc]="user"
>  fi
>done
>
>for user in "${relaylist[@]}"
>do
>  userenc=$(echo "$user"|base32 | tr = 9)
>  if [[ ! ${permissions[$userenc]+_} ]] # user doesn't exist
>  then
>    permissions[$userenc]="relay"
>  fi
>done
>
>permissionstr="permissions:"
>for userenc in "${!permissions[@]}"
>do
>  user=$(echo $userenc|tr 9 =|base32 -d)
>  permissionstr=$(printf "%s\n    '%s': %s" "$permissionstr" "$user" "${permissions[$userenc]})
>done
>```
>
>At this point, you have a correct permission string, but what to do? Store it in yunohost's app config file? It would work, but then it will be hard to read it again and parse it into lists. Maybe save both `permissionstr` and the individual lists, and only rely on `permissionstr` for the template.
>
>I feel like setting management in yunohost is extremely broken, which is a shame. These scripts are getting very complex already, so I feel like there is little value in keeping them written in bash and not, say, Python.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/YunoHost-Apps/mautrix_signal_ynh/pull/70#issuecomment-1379242366
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- Sent from /e/ Mail.

**Comment by nathanael-h** [comment]
Just saw that bridge name is weired. I saw in https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/74#issuecomment-1377178866 that you suggest to remove this setting from config panel and let customization only at install time: this seems ok for me. 
image


`Signal bridge bot a changé son nom d’affichage en Signal;bridge;bot`

**Comment by Gredin67** [comment]
@nathanael-h https://github.com/nathanael-h/mautrix_signal_ynh/pull/2
this is the up-to-date branch, that you should merge into yours if it works!

**Comment by nathanael-h** [comment]
I found some bugs in generated config file : 


```
Jan 12 11:22:09 aa.fr python3[3585079]: ruamel.yaml.parser.ParserError: while parsing a block mapping
Jan 12 11:22:09 aa.fr python3[3585079]:   in "config.yaml", line 115, column 5
Jan 12 11:22:09 aa.fr python3[3585079]: expected <block end>, but found '<block mapping start>'
Jan 12 11:22:09 aa.fr python3[3585079]:   in "config.yaml", line 274, column 9
```

L115 : 
```yaml
     username_template: sg_{userid}`
```
missing single quotes around the value I think

L271-274 :
```yaml
271     relay: true
272         # Whether relay mode should be allowed. If allowed, `!signal set-relay` can be used to turn any
273         # authenticated user into a relaybot for that chat.
274         enabled: true
```
Do you had this issues? I did not already look for fixing them. 


About https://github.com/nathanael-h/mautrix_signal_ynh/pull/2 : didn't test yet.

**Comment by Gredin67** [comment]
I just told you that you should test and merge https://github.com/nathanael-h/mautrix_signal_ynh/pull/2
not this PR here ;)

**Comment by nathanael-h** [comment]
Thanks @Gredin67 sorry I did not understood what you said before, I merged your commits here, and granted you write on this PR/repo. I tested what was not working before: it's working! 
I remove WIP status, and run CI-check.

**Comment by nathanael-h** [comment]
!test-me

**Comment by nathanael-h** [comment]
!testme

**Comment by Gredin67** [comment]
!testme

**Comment by Gredin67** [comment]
@nathanael-h could you "sync" you branch with the https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/74 (structure of the panel reworked, etc.)

**Comment by Gredin67** [comment]
@YunoHost-Apps/matrix-bridges we finally got the whatsapp config panel released, feel free to sync with you bridges!

**Comment by nathanael-h** [comment]
Wouhou !!! 🎉 👏👏👏👏 I'll try to catch up!

**Comment by Gredin67** [comment]
up @nathanael-h 

> @YunoHost-Apps/matrix-bridges we finally got the whatsapp config panel released, feel free to sync with you bridges!

**Comment by nathanael-h** [comment]
Sorry, I did not managed to free time to finish this PR... If someone would like to do it, I can grant write permission on my fork for easy contrib... thanks for the kind reminder!

**Comment by nathanael-h** [comment]
I reworked config panel to be in sync with https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/ 
WARNING: Not tested yet. 

!test-me

**Comment by nathanael-h** [comment]
!test-me

**Comment by Gredin67** [comment]
!testme

@nathanael-h it looks like you did not upgrade config.yaml to fit example-config.yaml from 0.4.3, see :
[mautrix_signal/example-config.yaml](https://github.com/mautrix/signal/compare/v0.4.2...v0.4.3#diff-4dbf0472836f12f108cc0d26103750a3e378ead8ec812c10e89c80fccd801610)

Then we need to adapt the __FOOBAR__ parameters accordingly.

Thanks for continuing this :dancers:

**Comment by nathanael-h** [comment]
Thanks for the remind! PR here targeting testing, and quickly master https://github.com/YunoHost-Apps/mautrix_signal_ynh/pull/83 @Gredin67 could you test it on your server? Because, actually mine is using the [config panel branch](https://github.com/nathanael-h/mautrix_signal_ynh/tree/config_panel)

I will also update this branch and the config panel to match, here, on this PR.

**Comment by nathanael-h** [comment]
This PR needs at least a fix for the permission system. The getter and setter are not working for permissions. Probably because this bridges uses simple quotes and not double quotes :
- https://github.com/nathanael-h/mautrix_signal_ynh/blame/19a63d024fac7ef003d07840aeb686be03cce59d/conf/config.yaml#L307
- https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/blob/53f02f8d80cf3c9491420a0f612afbbae86e8ff0/conf/config.yaml#L415

And more testing of everything else. 

> Then we need to adapt the FOOBAR parameters accordingly. 

I don't get it sorry. What are you refering to? Maybe the diff from 0.4.2 to 0.4.3? No key used by config panel was changed.

**Comment by Gredin67** [comment]
> The getter and setter are not working for permissions. Probably because this bridges uses simple quotes and not double quotes :
> 
>     * https://github.com/nathanael-h/mautrix_signal_ynh/blame/19a63d024fac7ef003d07840aeb686be03cce59d/conf/config.yaml#L307
> 
>     * https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/blob/53f02f8d80cf3c9491420a0f612afbbae86e8ff0/conf/config.yaml#L415

@rosbeef can you have a look, it should be super easy for you. Or maybe @thardev ?

**Comment by Gredin67** [comment]
!testme

**Comment by Gredin67** [comment]
> The getter and setter are not working for permissions. Probably because this bridges uses simple quotes and not double quotes :
> 
>     * https://github.com/nathanael-h/mautrix_signal_ynh/blame/19a63d024fac7ef003d07840aeb686be03cce59d/conf/config.yaml#L307
> 
>     * https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/blob/53f02f8d80cf3c9491420a0f612afbbae86e8ff0/conf/config.yaml#L415

@rosbeef @thardev can you have a look at this single/double quote sed issue ?

**Comment by rosbeef** [comment]
i did something on nathanael-h git fork
as soon as he merge you could test.
personnaly i have tested on my own fork and it works :
[x] config panel open
[x] settings are saved
[x] config file is modified
[x] settings are retreived

optionally i had to make the rustup modification to install my fork on my armhf platforme please push to inclue it on the master branch  ;)

**Comment by Gredin67** [comment]
@rosbeef if you find time, could you open a PR from your branch  https://github.com/rosbeef/mautrix_signal_ynh/tree/patch-3 to the testing branch of the main repo ? 
I just merged #83 and we only have to make sure that the config panel PR for signal matches [the one for mautrix_whatsapp](https://github.com/YunoHost-Apps/mautrix_whatsapp_ynh/pull/114) which should work properly !

**Comment by nathanael-h** [comment]
I updated this branch with @rosbeef modifications and merged testing in it. I'll grant you permission to push directly on it so that you can move without waiting for me.

**Comment by nathanael-h** [comment]
I tried the lists for relay, admin and users, had some issues (duplicate keys in the yaml file, no quotes around strings with * or @) and tried to provide fixes. It seems to work. 
I still have an issue with the main.bridge.enable_relaybot setting. I work in cli, but in webadmin I can only disable, enabling do not work. I don't understand why.

**Comment by nathanael-h** [comment]
This PR is tool old, I close it. If I find time I'd try a new one based on the whatsapp bridge. I someone else wants to do it, please feel free to do :smiley_cat:

**Pull request closed without merging**